fuel-core/src/database/resource.rs

```diff
@@ -126,8 +127,13 @@ impl<'a> AssetQuery<'a> {
             .map(|results| Ok(results??))
             .filter_ok(|coin| {
                 if let Resource::Coin { fields, .. } = coin {
-                    fields.asset_id == self.asset.id
-                        && fields.status == CoinStatus::Unspent
+                    let allowed = if let Some(assets) = &self.assets {
+                        assets.contains(&fields.asset_id)
+                    } else {
+                        true
+                    };
+
+                    allowed && fields.status == CoinStatus::Unspent
```

Comment:
I'd turn this upside down a bit:

```suggestion
    let is_unspent = fields.status == CoinStatus::Unspent;

    self.assets
        .as_ref()
        .map(|assets| assets.contains(&fields.asset_id) && is_unspent)
        .unwrap_or(is_unspent)
```